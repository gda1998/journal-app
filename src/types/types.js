const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] SetError',
    uiRemoveError: '[UI] RemoveError',
    uiStartLoading: '[UI] StartLoading',
    uiFinishLoading: '[UI] FinishLoading',

    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load notes',
    notesUpdate: '[Notes] Update note',
    notesFileUrl: '[Notes] Update note image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning'
};

export default types;