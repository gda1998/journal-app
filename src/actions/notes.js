import { db } from '../firebase/firebaseConfig';
import types from '../types/types';
import {
    addDoc,
    collection,
    doc,
    updateDoc,
    deleteDoc,
    // getDoc
} from 'firebase/firestore';
import Swal from 'sweetalert2';
import { swalAlert, swalLoading } from '../helpers/swalToast';
import loadNotes from '../helpers/loadNotes';
import fileUpload from '../helpers/fileUpload';
// import fileDelete from '../helpers/fileDelete';

export const startNewNoteAction = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const notesRef = collection(db, `${uid}/journal/notes`);
        return addDoc(notesRef, newNote)
            .then(docRef => {
                dispatch(activeNoteAction(docRef.id, newNote));
                dispatch(addNewNoteAction(docRef.id, newNote));
            })
            .catch(console.error);
    }
}

export const activeNoteAction = (id, note) => ({
    type: types.notesActive,
    payload: { id, ...note }
});

export const addNewNoteAction = (id, note) => ({
    type: types.notesAddNew,
    payload: { id, ...note }
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotesAction(notes));
    }
}

export const setNotesAction = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startUpdateNote = (note) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        !note.url && (note.url = null);
        const noteRef = doc(db, `${uid}/journal/notes/`, note.id);
        updateDoc(noteRef, note);
        dispatch(refreshNoteAction(note));
        swalAlert('Updated!', 'The note was updated', 'success');
    }
}

export const refreshNoteAction = (note) => ({
    type: types.notesUpdate,
    payload: note
});

export const startUploadingImgAction = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;
        swalLoading('Uploading', 'Please wait...');

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        Swal.close();
        await dispatch(startUpdateNote(activeNote));
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        try {
            const ref = await doc(db, uid, `journal/notes/${id}`);
            // const noteRef = await getDoc(ref);

            // Delete the image from cloudinary and from the db
            // fileDelete(noteRef.data().url)
            await deleteDoc(ref);

            dispatch(deleteNote(id));
        } catch (error) {
            console.error('Error al eliminar la nota: ', error);
        }
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})