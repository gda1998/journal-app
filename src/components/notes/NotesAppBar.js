import { useDispatch, useSelector } from 'react-redux';
import { startUpdateNote, startUploadingImgAction } from '../../actions/notes';

export const NotesAppBar = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );
    const handleUpdate = () => dispatch( startUpdateNote(note) );
    const handlePictureClick = () => document.querySelector('#fileSelector').click();
    
    const handleFileChange = ({target}) => {
        const file = target.files[0];
        file && dispatch( startUploadingImgAction(file) );
    }

    return (
        <div className="notes__appbar">
            <span>05 de Noviembre de 2021</span>

            <input
                id="fileSelector"
                onChange ={ handleFileChange }
                type="file"
                style={{ display: 'none' }}
            />

            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={ handleUpdate }
                >
                    Save
                </button>
            </div>
        </div>/* /.notes__appbar */
    );
}