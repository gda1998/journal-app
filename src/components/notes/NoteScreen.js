import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';
import { activeNoteAction, startDeleteNote } from '../../actions/notes';

export const NoteScreen = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );

    // useForm
    const [ formValues, handleInputChange, reset ] = useForm(note);
    const { title, body } = formValues;
    note.url && ( formValues.url = note.url );

    // useRef
    const activeId = useRef(note.id);

    // handleDelete
    const handleDelete = () => dispatch( startDeleteNote(note.id) );

    useEffect(() => {
        if( activeId.current !== note.id ){
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch( activeNoteAction(formValues.id, {...formValues}) );
    }, [dispatch, formValues]);
    

    return (
        <div className="notes__main-content">

            <NotesAppBar />
            <div className="notes__content">

                <input
                    autoComplete="off"
                    className="notes__title-input"
                    name="title"
                    onChange={ handleInputChange }
                    placeholder="Some awesome title"
                    type="text"
                    value={ title }
                />
                <textarea
                    className="notes__textarea"
                    name="body"
                    onChange={ handleInputChange }
                    placeholder="What's happened today?"
                    type="text"
                    value={ body }
                ></textarea>

                {
                    note.url &&
                    <div className="notes__image">
                        <img
                            src={ note.url }
                            alt="imagen"
                        />
                    </div>/* /.notes__image */
                }

            </div>{/* /.notes__content */}

            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>

        </div>/* /.notes__main-content */
    );
}
