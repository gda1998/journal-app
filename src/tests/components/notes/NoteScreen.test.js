import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import types from '../../../types/types';

const mockStore = configureStore([thunk]);
const initialStore = {
    ui: { loading: false, msgError: null },
    notes: {
        active: { 
            id: '1234',
            title: 'I am a title',
            body: 'I am the note body',
            url: 'https://any.com'
        }
    }
};
const store = mockStore(initialStore);

describe('Pruebas en <NoteScreen />', () => { 
    let wrapper;
    const getWrapper = (store) => mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>);

    beforeEach( () => {
        wrapper = getWrapper(store);
        store.clearActions();
    });

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar el activeNoteAction', () => { 
        wrapper.find('input[type="text"]').simulate('change', {
            target: { name: 'title', value: 'Hola Mundo' }
        });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.notesActive,
            payload: {             
                id: '1234',
                title: 'Hola Mundo',
                body: 'I am the note body',
                url: 'https://any.com'
            }
        });
    });
});