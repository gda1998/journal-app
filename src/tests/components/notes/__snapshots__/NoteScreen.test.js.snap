// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> Debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      05 de Noviembre de 2021
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="I am a title"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What's happened today?"
      type="text"
      value="I am the note body"
    />
    <div
      className="notes__image"
    >
      <img
        alt="imagen"
        src="https://any.com"
      />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
